pkg/volume/nfs/nfs.go

```diff
@@ -122,9 +124,12 @@ func (nfsVolume *nfs) SetUpAt(dir string) error {
 	if mountpoint {
 		return nil
 	}
-	exportDir := nfsVolume.exportPath
 	os.MkdirAll(dir, 0750)
-	err = nfsVolume.mounter.Mount(nfsVolume.server, exportDir, dir, nfsVolume.readOnly)
+	flags := uintptr(0)
+	if nfsVolume.readOnly {
+		flags = mount.FlagReadOnly
+	}
+	err = nfsVolume.mounter.Mount(fmt.Sprintf("%s:%s", nfsVolume.server, nfsVolume.exportPath), dir, "nfs", flags, "")
```

Comment:
What's wrong with NFS using its own mounter?  It's an interface for a
reason.

That said, I could maybe buy that we just implement it once in terms or
exec, but that should be a PR of its own so we can consider the
implications.

On Wed, Apr 1, 2015 at 6:59 PM, Deyuan Deng <notifications@github.com>
wrote:

> In pkg/volume/nfs/nfs.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6269#discussion_r27628120>
> :
>
> >  	os.MkdirAll(dir, 0750)
> > -	err = nfsVolume.mounter.Mount(nfsVolume.server, exportDir, dir, nfsVolume.readOnly)
> > +	flags := uintptr(0)
> > +	if nfsVolume.readOnly {
> > +		flags = mount.FlagReadOnly
> > +	}
> > +	err = nfsVolume.mounter.Mount(fmt.Sprintf("%s:%s", nfsVolume.server, nfsVolume.exportPath), dir, "nfs", flags, "")
>
> Whoops, forgot to push new changes. The code won't work, I tested with the
> following code
>
> Run a nfs server, expose container's '/lib' directory, running at 172.17.0.35
> $ docker run --name nfs --privileged cpuguy83/nfs-server /lib
>
> go code:
> if err := syscall.Mount("172.17.0.35:/lib", "/tmp/lib", "nfs", uintptr(0), ""); err != nil {
>     fmt.Println(err)
> }
>
> It complains 'invalid argument'. But mount command works. Out of
> curisity, I digged through some linux utils. mount command (in
> util-linux) uses /sbin/mount.nfs, which comes from nfs-utils. In nfs-utils,
> it passes some data to kernal
>
> http://git.linux-nfs.org/?p=steved/nfs-utils.git;a=blob;f=utils/mount/nfsmount.c;h=930622d94b57f6e4dd12497fb99277dc76580ecf;hb=HEAD#l858
>
> Inspired by the code, I tried to supply "addr=" option to mount syscall,
> and it works
>
> go code
> syscall.Mount("172.17.0.35:/lib", "/tmp/lib", "nfs", uintptr(0), "addr=172.17.0.35")
>
> $ ls /tmp/lib
> ifupdown  init  klibc-P2s_k-gf23VtrGgO2_4pGkQgwMY.so  lsb  modprobe.d  plymouth  resolvconf  systemd  terminfo  udev  x86_64-linux-gnu
>
> The tools exist for reasons, even if this works, I think we need to take
> advantage of them. But having nfs implements its own mounter doesn't sound
> right. How about we change linux mounter implementation to use
> exec('mount') instead of syscall.Mount?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6269/files#r27628120>
> .
>